swarm/storage/mru/resource.go

```diff
@@ -719,78 +756,49 @@ func (h *Handler) update(ctx context.Context, addr storage.Address, data []byte,
 	// an update can be only one chunk long; data length less header and signature data
 	// 12 = length of header and data length fields (2xuint16) plus period and frequency value fields (2xuint32)
 	datalimit := h.chunkSize() - int64(signatureLength-len(rsrc.name)-12)
-	if int64(len(data)) > datalimit {
-		return nil, NewError(ErrDataOverflow, fmt.Sprintf("Data overflow: %d / %d bytes", len(data), datalimit))
-	}
-
-	// get our blockheight at this time and the next block of the update period
-	currentblock, err := h.getBlock(ctx, rsrc.name)
-	if err != nil {
-		return nil, NewError(ErrIO, fmt.Sprintf("Could not get block height: %v", err))
-	}
-	nextperiod, err := getNextPeriod(rsrc.startBlock, currentblock, rsrc.frequency)
-	if err != nil {
-		return nil, err
+	if int64(len(mru.data)) > datalimit {
+		return nil, NewError(ErrDataOverflow, fmt.Sprintf("Data overflow: %d / %d bytes", len(mru.data), datalimit))
 	}
 
-	// if we already have an update for this block then increment version
-	// resource object MUST be in sync for version to be correct, but we checked this earlier in the method already
-	var version uint32
-	if h.hasUpdate(addr, nextperiod) {
-		version = rsrc.version
-	}
-	version++
-
-	// calculate the chunk key
-	key := h.resourceHash(nextperiod, version, rsrc.nameHash, rsrc.owner)
-
-	// if we have a signing function, sign the update
-	// \TODO this code should probably be consolidated with corresponding code in New()
-	var signature *Signature
-	if h.signer != nil {
-		// sign the data hash with the key
-		digest := h.keyDataHash(key, data)
-		sig, err := h.signer.Sign(digest)
-		if err != nil {
-			return nil, NewError(ErrInvalidSignature, fmt.Sprintf("Sign fail: %v", err))
+	if rsrc.period == mru.period {
+		if mru.version != rsrc.version+1 {
+			return nil, NewError(ErrInvalidValue, fmt.Sprintf("Invalid version for this period. Expected version=%d", rsrc.version+1))
 		}
-		signature = &sig
-
-		// get the address of the signer (which also checks that it's a valid signature)
-		_, err = getAddressFromDataSig(digest, *signature)
-		if err != nil {
-			return nil, NewError(ErrInvalidSignature, fmt.Sprintf("Invalid data/signature: %v", err))
+	} else {
+		if !(mru.period > rsrc.period && mru.version == 1) {
+			return nil, NewError(ErrInvalidValue, fmt.Sprintf("Invalid version,period. Expected version=1 and period > %d", rsrc.period))
```

Comment:
This check here is key to make sure we are creating an update that is ahead of what we already know exists

Reply:
If I remember correctly code resembling this was also used in the `LookupPrevious` method(s). Could we share it?

Reply:
Code in `LookupPrevious` doesn't look like this. What do you mean?

Reply:
Code in `LookupPrevious` does not look like this. What do you mean?

Reply:
yes right, `LookupPrevious` has it the other way around.